patches/api/0416-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,758 @@
++     * Registers the command builder into the specified dispatcher.
++     *
++     * @param dispatcher dispatcher
++     * @return if the command was registered and added to the command map in some form (e.g. no command collisions)
++     */
++    public abstract boolean register(@NotNull CommandDispatcher<CommandSourceStack> dispatcher);
++
++    @ApiStatus.Internal
++    private static class ProviderHolder {
++
++        private static final CommandBuilderProvider PROVIDER = Services.service(CommandBuilderProvider.class)
++            .orElseThrow();
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandBuilderProvider.java b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilderProvider.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..1f5a17fc682948baf83122285c6dfc8c7e8dbf16
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilderProvider.java
+@@ -0,0 +1,9 @@
++package io.papermc.paper.command.brigadier;
++
++import org.bukkit.plugin.Plugin;
++import org.jetbrains.annotations.ApiStatus;
++
++@ApiStatus.Internal
++public interface CommandBuilderProvider {
++    CommandBuilder create(Plugin plugin, String name);
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..fc88435a149918b2675d94b80d659ab0051f5541
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
+@@ -0,0 +1,65 @@
++package io.papermc.paper.command.brigadier;
++
++import com.mojang.brigadier.context.CommandContext;
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import io.papermc.paper.command.brigadier.argument.ArgumentResolver;
++import org.bukkit.Location;
++import org.bukkit.World;
++import org.bukkit.command.CommandSender;
++import org.bukkit.entity.Entity;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * Represents a vanilla command source stack which is used
++ * for command execution.
++ */
++@ApiStatus.NonExtendable
++@ApiStatus.Experimental
++public interface CommandSourceStack {
++
++    /**
++     * Gets the bukkit entity running this command.
++     * @return bukkit entity, null if not present
++     */
++    @Nullable
++    Entity getBukkitEntity();
++
++    /**
++     * Gets the world that this command is being executed in.
++     * @return world
++     */
++    @Nullable
++    World getBukkitWorld();
++
++    /**
++     * Gets the location that this command is being executed in.
++     * @return location
++     */
++    @Nullable
++    Location getBukkitLocation();
++
++    /**
++     * Gets the sender representing this source stack.
++     * @return source stack sender
++     */
++    @NotNull
++    CommandSender getBukkitSender();
```

Comment:
I don't really like these names. If they're required for compat with the existing brig api, we should deprecate and give them actual names, "bukkit" shouldn't be in the name, just `getEntity`, `getWorld`, etc.